enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,446 @@
+  - "eth0"
+
+---
+```
+
+- sample `IngressNodeFirewall` object
+
+```yaml
+kind: IngressNodeFirewall
+metadata:
+  name: worker-nodes-firewall-on-eth0
+  annotations:
+    node-endpoint: worker-eth0
+spec:
+  nodeSelector: has(do-node-ingress-firewall)
+  ingress:
+  - fromCIDRS:
+       - 1.1.1.1/24
+       - 100:1::1/64
+    rules:
+    - order: 10  
+      protocol: tcp
+      ports: [100]
+      action: allow
+    - order: 20
+      protocol: udp
+      ports: [8000, 8080, 9000]
+      action: allow
+  - fromCIDRS:
+       - 3.3.3.3/24
+    rules:
+    - order: 10
+      protocol: icmp
+      type: 3 # Destination unreachable
+      code: 1 # Host unreachable
+      action: allow
+  - fromCIDRS:
+       - 10:10::1/64
+    rules:
+    - order: 10
+      protocol: icmpv6
+      type: 128 # Ping request
+      action: deny
+  - fromCIDRS:
+       - 0.0.0.0/0
+       - 0::0/0
+    rules:
+    - order: 10
+      protocol: tcp
+      action: allow
+
+---
+
+kind: IngressNodeFirewall
+metadata:
+  name: worker-nodes-firewall
+  annotations:
+    node-endpoint: worker-all-interfaces
+spec:
+  nodeSelector: has(do-node-ingress-firewall)
+  ingress:
+  - fromCIDRS:
+       - 1.1.1.1/24
+       - 100:1::1/64
+    rules:
+    - order: 30
+      protocol: udp
+      ports: 8000-9000
+      action: deny
+  - fromCIDRS:
+       - 0.0.0.0/0
+       - 0::0/0
+    rules:
+    - order: 20
+      protocol: tcp
+      action: deny
+    - order: 30
+      protocol: udp
+      action: deny
+    - order: 40
+      protocol: sctp
+      action: deny
+```
+
+In case of multiple `IngressNodeFirewall` on the same nodes and interfaces if the CIDR matches or overlap the `order` field will specify which rule applied first, lower order will be excuted first, once the rule is excuted and action is taken firewall operation will be completed, all the following rules won't be checked. 
+
+- eBPF data structures and types
+
+```c
+#define DENY XDP_DROP
+#define ALLOW XDP_PASS
+#define MAX_DST_PORTS 100
+struct ruleInfo {
+    __u32 ruleId;
+    __u8 protocol;
+    __u32 srcAddr;
+    __u32 srcMask;
+     __u16 dstPort[MAX_DST_PORTS];
+    __u8 icmpType;
+    __u8 icmpCode;
+    __u8 action;
+} __packed;
+
+#define MAX_INSIDE_TARGETS   (1024)
+#define MAX_RULES_PER_TARGET (100)
+
+struct bpf_map_def SEC("maps") ingressRulesTbl = {
+    .type = BPF_MAP_TYPE_LPM_TRIE,
+    .key_size = sizeof(struct bpf_lpm_ip_trie),
+    .value_size = sizeof(struct rulesStore) + (sizeof(struct ruleInfo)*MAX_RULES_PER_TARGET),
+    .max_entries = MAX_INSIDE_TARGETS,
+    .map_flags = BPF_F_NO_PREALLOC,
+};
+
+// using Longest perfix match in case of overlapping CIDRs we need to match to the more specific CIDR.
+struct bpf_lpm_trie_ip_key {
+    __u32 prefixlen;
+    union {
+        __u8 data[4];
+        __u8 data[16];
+    } u;
+} __packed;
+
+struct rulesStore {
+    __u32 numRules;
+    struct ruleInfo rulesInfo[0];
+} __packed;
+```
+
+![eBPF datastructures](./ebpf-structures.png)
+
```

Comment:
You mention updating stats earlier, but I don't see any design details on how you plan to manage:

1. Updating stats from eBPF (which will need to be done percpu to avoid synchronization issues)
2. Counter behaviour - is it going to wrap around when you hit u64::MAX?
2. Reading map counters into userspace: assume polling?